### PR TITLE
Search

### PR DESCRIPTION
**Pull request by silkspace**
Adds search + mlb encoder

### CODE REVIEW

**Comment by silkspace** [comment]
@lmeyerov let me know if there is anything that is needed

**Review by lmeyerov — CHANGES_REQUESTED** [review]
Super close! See inline comments!

- [ ] some missing docs
- [ ] a couple non-functional interfaces / unsafe memoization
- [ ] multilabel: maybe default on (opt-in to any col with lists? or not b/c it's slow?), and when it is opted-in, allow picking which cols vs all
- [ ] various prints that should be disabled / logger

File: graphistry/feature_utils.py [hunk 1]
```diff
@@ -2117,13 +2211,14 @@ def featurize(
...
-        use_scaler: Optional[str] = "minmax",
-        use_scaler_target: Optional[str] = "kbins",
+        use_scaler: Optional[str] = None,
+        use_scaler_target: Optional[str] = None,
         cardinality_threshold: int = 40,
         cardinality_threshold_target: int = 400,
         n_topics: int = config.N_TOPICS_DEFAULT,
         n_topics_target: int = config.N_TOPICS_TARGET_DEFAULT,
-        embedding=False,
+        multilabel: bool = False,
```
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/feature_utils.py:2220]: `multilabel` works for a column that is a list of lists, so by default, should be False. 
so for example `df = pd.DataFrame({'col': ["a, b, c", "b, c", "d"])`

```
df['mlb'] = df.col.apply(lambda x: x.split(', '))
g = graphistry.nodes(df)
g2 = g.umap(X=..., y=['mlb'], multilabel=True)
```

File: graphistry/constants.py [hunk 2]
```diff
@@ -1,19 +1,19 @@
 # ###############################################################
-VERBOSE = None  # set to true for info, false for debug, None for none
+VERBOSE = True  # set to true for info, false for debug, None for none
```
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/constants.py:2]: oof yes

File: graphistry/feature_utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/feature_utils.py:2220]: In fact, perhaps I should add an `Exception` if `multilabel=True` on a column that is not list of list (currently would just error ungracefully)

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,13 @@ The changelog format is based on [Keep a Changelog](https://keepachangelog.com/e
 This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and all PyGraphistry-specific breaking changes are explictly noted here.
 
 ## [Development]
+## [0.28.1 - 2022-10-10]
+
+### Added
+
+* AI: full text & semantic search (`g.query(..)` and `g.query_graph(..).plot()`)
+* Featurization: support for dataframe columns that are list of lists -> multioutput targets
```
[the remark this answers is not in the document]
Comment (silkspace) [reply CHANGELOG.md:12]: yes but only if we unpin dirty_cat and change it to v0.3.0

File: graphistry/text_utils.py [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        results[DISTANCE] = distances
+        results = results.query(f"{DISTANCE} < {thresh}")
+
+        results = results.sort_values(by=[DISTANCE])
+
+        return results, vect
+    
+    def _query(self, query: str, top_k: int, thresh: float):
+        # build the query dataframe
+        if not hasattr(self, 'search_index'):
```
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/text_utils.py:95]: Fixed

File: graphistry/text_utils.py [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+                    "uint16",
+                ]:
+                        qdf[col] = df[col].mean()
+
+        return self._query_from_dataframe(qdf, thresh=thresh, top_k=top_k)
+
+    def search(
+        self, query: str, cols = None, thresh: float = 5000, fuzzy: bool = True, top_k: int = 10
+    ):  
+        """NL-query and return dataframe of results
```
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/text_utils.py:144]: Done, thanks

[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/text_utils.py:144]: I added it in the main `def search(..)` function

File: graphistry/text_utils.py [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    def search_graph(
+        self,
+        query: str,
+        scale: float = 0.5,
+        top_k: int = 100,
+        thresh: float = 5000,
+        broader: bool = False,
+        inplace: bool = False,
+    ):
+        """Input a natural language query and return a graph of
```
[the remark this answers is not in the document]
Comment (silkspace) [reply graphistry/text_utils.py:184]: Done

**Pull request merged**